File: Problems/8. String to Integer (atoi)/string-to-integer-atoi.dart
//Solution: 238ms 143.10 MB
class Solution {
  int myAtoi(String s) {
    s = s.trim();
    if (s.isEmpty) {
      return 0;
    }

    int result = 0;
    int sign = 1;
    int i = 0;

    // Handle sign
    if (s[i] == '-' || s[i] == '+') {
      sign = (s[i++] == '-') ? -1 : 1;
    }

    // Process digits
    while (i < s.length && s[i].compareTo('0') >= 0 && s[i].compareTo('9') <= 0) {
      int digit = int.parse(s[i]);
      
      // Check for overflow during digit processing
      if (result > (2147483647 - digit) ~/ 10) {
        return (sign == 1) ? 2147483647 : -2147483648;
      }

      result = result * 10 + digit;
      i++;
    }

    // Apply sign
    return result * sign;
  }
}

File: Problems/8. String to Integer (atoi)/string-to-integer-atoi.ts
//Solution: 54ms 45.74 MB
function myAtoi(s: string): number {
  const bound = 2 ** 31 - 1;
  s = s.trim();

  let sign = 1;
  let itr = 0;
  let result = 0;

  if (s[itr] === "-" || s[itr] === "+") {
    sign = s[itr++] === "-" ? -1 : 1;
  }

  while (s[itr] >= "0" && s[itr] <= "9") {
    const digit = parseInt(s[itr], 10);

    if (result > (bound - digit) / 10) {
      return sign === 1 ? bound : -bound - 1;
    }

    result = result * 10 + digit;
    itr++;
  }

  return sign * result;
}

//Solution 2: 44ms 44.90MB
const clamp = (num: number, min: number, max: number): number => {
  return Math.max(Math.min(num, max), min);
};

const myAtoi = (s: string): number => {
  const parsedNumber = parseInt(s.trim());

  if (Number.isNaN(parsedNumber)) {
    return 0;
  }

  return clamp(parsedNumber, -2147483648, 2147483647);
};
